src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/RemoteTagHelperDeltaProvider.cs

```diff
@@ -3,37 +3,33 @@
 
 using System.Collections.Immutable;
 using Microsoft.AspNetCore.Razor.Language;
-using Microsoft.AspNetCore.Razor.PooledObjects;
 using Microsoft.AspNetCore.Razor.Serialization;
+using Microsoft.AspNetCore.Razor.Utilities;
 
 namespace Microsoft.CodeAnalysis.Remote.Razor;
 
 internal class RemoteTagHelperDeltaProvider
 {
-    private readonly TagHelperResultCache _resultCache;
-    private readonly object _resultIdLock = new();
+    private readonly TagHelperResultCache _resultCache = new();
+    private readonly object _gate = new();
     private int _currentResultId;
 
-    public RemoteTagHelperDeltaProvider()
-    {
-        _resultCache = new TagHelperResultCache();
-    }
-
     public TagHelperDeltaResult GetTagHelpersDelta(
         ProjectId projectId,
         int lastResultId,
         ImmutableArray<TagHelperDescriptor> currentTagHelpers)
     {
         var cacheHit = _resultCache.TryGet(projectId, lastResultId, out var cachedTagHelpers);
+
         if (!cacheHit)
         {
             cachedTagHelpers = ImmutableArray<TagHelperDescriptor>.Empty;
         }
 
-        var added = GetAddedTagHelpers(currentTagHelpers, cachedTagHelpers!);
-        var removed = GetRemovedTagHelpers(currentTagHelpers, cachedTagHelpers!);
+        var added = TagHelperDelta.Compute(cachedTagHelpers, currentTagHelpers);
+        var removed = TagHelperDelta.Compute(currentTagHelpers, cachedTagHelpers);
```

Comment:
It's great that this loops over better data structures, but seems like it has just as much looping, if not a little bit more. Is it worth investigating if it's faster to sort the `currentTagHelpers` (we can assume `cachedTagHelpers` will already be sorted, since it was previously a `currentTagHelpers`) and then process both lists in one loop looking for additions and removals?

Reply:
I didn't _quite_ understand the approach you described. However, the change here has considerably less looping than before. Previously, delta computation was O(n^2) because `ImmutableArray<TagHelperDescriptor>.Contains(...)` was called inside the loop over the second array. Now there's just two loops -- one to fill a hash set with the first array (fast) and one to loop over the second array and check the hash set (fast). Sorting would result in lots of up-front comparisons and look-up on the sorted list would be O(log N) vs. O(1) to look-up in a hash set. I'm not seeing how that would go further, but maybe I'm missing something?

Regardless, I think this improvement is fast enough. :smile:

Reply:
I agree this is fast enough, for sure.

What I was thinking though, is that if you had two lists, both sorted (and yes, there is an up front cost for sorting one of them), then you could loop through both lists at the same time, comparing items from the left and right lists to determine which things has been added and removed. Whether that is a net benefit over creating sets and looking things up in them, I don't know.

Reply:
Got it -- thanks for the extra explanation! Yes, we would have to sort both lists, which would likely be costlier than adding them to a hash set because sorting will generally require comparing items many more times than inserting them into a hash table. And, look up on a sorted list is O(log N) for a binary search vs. O(1) in a hash set. It would definitely work, but I would expect the difference between that approach and this one to be negligible or slightly slower due to the up-front cost.